packages/react-reconciler/src/ReactFiberHooks.new.js

```diff
@@ -1706,6 +1711,71 @@ function rerenderOpaqueIdentifier(): OpaqueIDType | void {
   return id;
 }
 
+function mountRefresh() {
+  // TODO: CacheInstance should never be null. Update type.
+  const cacheInstance: CacheInstance | null = readContext(CacheContext);
+  return mountCallback(refreshCache.bind(null, cacheInstance), [cacheInstance]);
+}
+
+function updateRefresh() {
+  const cacheInstance: CacheInstance | null = readContext(CacheContext);
+  return updateCallback(refreshCache.bind(null, cacheInstance), [
+    cacheInstance,
+  ]);
+}
+
+function refreshCache<T>(
+  cacheInstance: CacheInstance | null,
+  seedKey: ?() => T,
+  seedValue: T,
+) {
+  if (cacheInstance !== null) {
+    const provider = cacheInstance.provider;
+
+    // Inlined startTransition
+    // TODO: Maybe we shouldn't automatically give this transition priority. Are
+    // there valid use cases for a high-pri refresh? Like if the content is
+    // super stale and you want to immediately hide it.
+    const prevTransition = ReactCurrentBatchConfig.transition;
+    ReactCurrentBatchConfig.transition = 1;
+    // TODO: Do we really need the try/finally? I don't think any of these
+    // functions would ever throw unless there's an internal error.
+    try {
+      const eventTime = requestEventTime();
+      const lane = requestUpdateLane(provider);
+      // TODO: Does Cache work in legacy mode? Should decide and write a test.
+      const root = scheduleUpdateOnFiber(provider, lane, eventTime);
+
+      let seededCache = null;
+      if (seedKey !== null && seedKey !== undefined && root !== null) {
+        // TODO: Warn if wrong type
+        seededCache = new Map([[seedKey, seedValue]]);
+        transferCacheToSpawnedLane(root, seededCache, lane);
```

Comment:
This seems sketchy because it's mutating the existing entry. What if this has already started rendering? It would mutate it while in the middle of rendering and potentially read inconsistent value. It would also override what's already there. So if you refresh two different roots with different seeds in the same transition, it would forget one.

Reply:
It doesn't overwrite, it will prefer an existing cache, but I agree with your overall point that for refreshes we can store it on the queue instead to avoid conflicts. I'll go ahead and change that.